sympy/polys/polytools.py

```diff
@@ -3816,6 +3816,72 @@ def nth_power_roots_poly(f, n):
 
         return r.replace(t, x)
 
+    def which_roots(f, candidates, real=True):
+        """
+        Given a superset of roots of f, finds which ones are
+        roots of f. Note this won't work properly if candidates is
+        not a superset of the roots of f. The list returned is the
+        same length as the number of roots of f, counted with
+        multiplicity, and preserves the order of the original candidates list.
+
+        Examples
+        ========
+
+        >>> from sympy import Poly, I
+        >>> from sympy.abc import x
+
+        >>> f = Poly(x**4 - 1)
+
+        >>> f.which_roots([-1, 1, 0, -2, 2], real=True)
+        [-1, 1]
+        >>> f.which_roots([-1, 1, -I, I, 0], real=False)
+        [-1, 1, -I, I]
+        >>> f.which_roots([2, 5], real=True) # expectedly wrong
+        [2, 5]
+        """
+        if f.is_multivariate:
+            raise MultivariatePolynomialError(
+                "Must be a univariate polynomial")
+
+        # must be QQ, ZZ, or Alg for proper counting
+        dom = f.get_domain()
+        if not (dom.is_ZZ or dom.is_QQ or dom.is_AlgebraicField):
+            raise NotImplementedError(
+                "root counting not supported over %s" % dom)
+
+        if real:
+            num_roots = f.count_roots()
+        else:
+            num_roots = f.degree()
+
+        prec = 10
+        # compare len(set()) bc expected behavior is for multiple roots
+        while len(set(candidates)) > num_roots:
+            candidates_filtered = []
+            processed = set() # track already seen
+
+            for c in candidates:
+                if c not in processed:
+                    r_f = f(c).evalf(prec, maxn=2*prec)
+                    processed.add(c)
+                    if abs(r_f)._prec < 2:
+                        candidates_filtered.append(c)
+                else:
+                    if c in candidates_filtered:
+                        # already approved this candidate
+                        candidates_filtered.append(c)
```

Comment:
I would write this as:
```python
root_counts = Counter(candidates)
roots_remaining = lambda: sum(root_counts.values())

while roots_remaining() > num_roots:
    for r in list(root_counts.keys()):
        # If f(r) != 0 then f(r).evalf() gives a float/complex with precision.
        f_r = f(r).evalf(prec, maxn=2*prec)
        if abs(f_r)._prec >= 2:
            root_counts.pop(c)
    prec *= 2

roots = sum([r]*c for r, c in root_counts.items(), [])
```

Reply:
I like this! I have changed the implementation to this. 

However, I found an unforeseen bug, which causes this implementation to loop indefinitely (until overflow of prec) -- and was also causing issues with the prior implementation. The issue is with multiplicity of real roots. This breaks real_roots() sometimes.

```
>>> Poly((x-1)**2, x).which_roots([1, 1])
OverflowError: cannot convert float infinity to integer

>>> Poly((x-sqrt(2))**2, x, extension=True).real_roots()
OverflowError: cannot convert float infinity to integer
```

The issue in the last example is that the lift has roots `[-sqrt(2), -sqrt(2), sqrt(2), sqrt(2)]`, the `-sqrt(2)` gets filtered out, but then roots_remaining() always stays 2, while num_roots is 1 -- this is because .count_roots() counts unique roots.

I think it is better to just compare the count of unique roots, but then add the multiplicity back in. This also leads to more consistent output between real=True and real=False (which uses degree to count roots).

I've added a function to Poly which is `root_multiplicity()` to accomplish this.

Reply:
I guess it is not easily possible to have `count_roots` return a count with multiplicities.

I think this means that we need to factorise the polynomial first. We could simply require in this method that the polynomial be square-free. It is not necessary to check that it is square-free but we can document the requirement and make sure it is satisfied before calling the function.

This goes back to what I said before about this being handled in the wrong place. The multiple roots case should be handled at a higher-level using factorisation and then at some lower level the problem to be solved only involves square free polynomials and multiple roots do not need to be considered.

Reply:
I see, that makes sense. I definitely think we need to handle multiplicity properly. I will work on the square-free stuff.